File: scripts/framework_audit.py
#!/usr/bin/env python3
"""
Content Framework Enforcer — kanokmiah.com.bd (v2)
Reads data.js, runs all 6 framework checks on each post.
More robust TF-IDF using tags + title-based keyword extraction.
"""
import re, json, sys
from pathlib import Path
from collections import Counter

DATA_FILE = Path("src/app/blog/data.js")
posts_raw = DATA_FILE.read_text(encoding="utf-8")

# --- Parse posts ---
posts = []
current = None
for line in posts_raw.splitlines():
    sm = re.search(r'slug:\s*"([^"]+)"', line)
    if sm:
        if current:
            posts.append(current)
        current = {"slug": sm.group(1), "raw": ""}
    if current:
        current["raw"] += line + "\n"
if current:
    posts.append(current)

def get_field(post_raw, field):
    """Extract a single-line or multi-line string field value."""
    m = re.search(rf'{field}:\s*"([^"]*)"', post_raw)
    if m:
        return m.group(1)
    m = re.search(rf'{field}:\s*\n\s*"([^"]*)"', post_raw)
    if m:
        return m.group(1)
    return ""

def get_tags_linear(post_raw):
    """Extract tags from the tags: [...] line."""
    # Find the tags line
    for line in post_raw.splitlines():
        line_s = line.strip()
        if line_s.startswith("tags:") or line_s.startswith("tags :"):
            # Extract all quoted strings
            return re.findall(r'"([^"]+)"', line_s)
    return []

def get_content(post_raw):
    """Extract content (backtick block) from post raw text."""
    m = re.search(r'content:\s*`(.*?)`\s*,?\s*\n', post_raw, re.DOTALL)
    if m:
        return m.group(1)
    idx = post_raw.find("content:")
    if idx < 0:
        return ""
    rest = post_raw[idx:]
    bt_start = rest.find("`")
    if bt_start < 0:
        return ""
    bt_end = rest.find("`,\n", bt_start + 1)
    if bt_end < 0:
        bt_end = rest.rfind("`")
        if bt_end <= bt_start:
            return ""
    return rest[bt_start+1:bt_end]

def extract_keywords(title, tags):
    """Extract primary keyword from title and tags intelligently."""
    # Priority 1: Use tags as keyword source if they're descriptive
    # Priority 2: Extract meaningful phrase from title
    
    # Common stop words
    stop = {"a", "an", "the", "for", "in", "of", "to", "your", "is", "are", 
            "why", "what", "how", "that", "this", "with", "and", "or", "but",
            "not", "all", "be", "have", "has", "do", "does", "its", "it",
            "at", "on", "by", "from", "as", "no", "yes", "so", "if", "about",
            "into", "over", "after", "before", "between", "under", "above",
            "below", "out", "off", "up", "down", "just", "more", "most",
            "also", "very", "can", "will", "just", "should", "need", "needs",
            "which", "who", "whom", "when", "where", "their", "them", "they",
            "was", "were", "been", "being", "has", "had", "have", "does",
            "did", "done", "get", "got", "getting"}
    
    # Clean title
    title_lower = title.lower().replace(":", "").replace("—", " ").replace("-", " ").replace("'", "").replace('"', "").replace("?", "").replace("!", "")
    title_words = [w for w in title_lower.split() if w not in stop and len(w) > 1]
    
    # Try using tags first (they're usually better keywords)
    if tags:
        # Pick the most relevant tag (not just "Bangladesh", "2026" etc.)
        generic_tags = {"bangladesh", "2026", "digital marketing", "seo", "bd"}
        specific_tags = [t for t in tags if t.lower() not in generic_tags and len(t) > 3]
        if specific_tags:
            # Use the first specific tag
            candidate = specific_tags[0].lower()
            # Check if candidate appears in content with word boundaries
            return candidate
    
    # Fall back to title
    if len(title_words) >= 3:
        return " ".join(title_words[:3])
    elif len(title_words) >= 2:
        return " ".join(title_words[:2])
    elif title_words:
        return title_words[0]
    return title_lower.strip()

def count_phrase(content_lower, phrase):
    """Count occurrences of phrase with word boundaries, handling hyphenated forms."""
    count = 0
    # Exact phrase
    count += len(re.findall(r'\b' + re.escape(phrase) + r'\b', content_lower))
    # With hyphen variations (e.g., "e commerce" -> "e-commerce")
    if " " in phrase:
        hyphenated = phrase.replace(" ", "-")
        count += len(re.findall(r'\b' + re.escape(hyphenated) + r'\b', content_lower))
    return count

def check_tfidf(post_raw, content):
    title = get_field(post_raw, "title")
    tags = get_tags_linear(post_raw)
    
    keyword = extract_keywords(title, tags)
    content_lower = content.lower()
    
    exact_count = count_phrase(content_lower, keyword)
    
    # For shorter keywords, also check individual words all appearing
    words = keyword.split()
    if len(words) >= 2:
        # Check each word appears sufficiently
        min_word_count = min(content_lower.count(w) for w in words)
    else:
        min_word_count = exact_count
    
    return {
        "keyword": keyword,
        "occurrences": exact_count,
        "pass": exact_count >= 5 or min_word_count >= 10  # either phrase appears enough or all words appear enough
    }

def check_entities(post_raw, content):
    content_lower = content.lower()
    
    entities_to_check = {
        "Location: Dhaka": bool(re.search(r'\bdhaka\b', content_lower)),
        "Country: Bangladesh": bool(re.search(r'\bbangladesh\b', content_lower)),
    }
    
    service_types = {
        "local seo": ["local seo", "google business profile", "google maps", "gbp"],
        "technical seo": ["technical seo", "core web vital", "page speed"],
        "link building": ["link building", "backlink"],
        "content marketing": ["content marketing"],
        "geo": ["generative engine optimization", " geo ", "ai search"],
        "aeo": ["answer engine optimization", " aeo "],
    }
    
    detected_services = set()
    for svc, keywords in service_types.items():
        for kw in keywords:
            if kw in content_lower:
                detected_services.add(svc)
    
    missing_entities = [k for k, v in entities_to_check.items() if not v]
    
    return {
        "detected_services": sorted(detected_services),
        "missing_entities": missing_entities,
        "pass": len(missing_entities) == 0
    }

def check_pillar_link(post_raw, content):
    content_lower = content.lower()
    pillar_pages = [
        "/blog/complete-seo-guide-bangladesh-businesses-2026",
        "/services",
        "/about",
    ]
    
    found_pillar_links = [pp for pp in pillar_pages if pp.lower() in content_lower]
    
    tags = get_tags_linear(post_raw)
    
    return {
        "tags": tags,
        "pillar_links": found_pillar_links,
        "pass": len(found_pillar_links) > 0
    }

def check_aeo_geo(post_raw, content):
    faq_questions = re.findall(r'^#{1,3}\s+.*\?', content, re.MULTILINE)
    return {
        "question_heading_count": len(faq_questions),
        "pass": len(faq_questions) >= 2
    }

def check_internal_links(post_raw, content):
    internal_links = re.findall(r'\[([^\]]+)\]\((/[^)]+)\)', content)
    meaningful = [(text, url) for text, url in internal_links 
                  if url != "/" and "https://kanokmiah.com.bd" not in url and "http" not in url]
    return {
        "total_internal_links": len(meaningful),
        "pass": len(meaningful) >= 3
    }

def check_schema(post_raw, content):
    title = get_field(post_raw, "title")
    excerpt = get_field(post_raw, "excerpt")
    date = get_field(post_raw, "date")
    author = get_field(post_raw, "author")
    
    missing = []
    if not title:
        missing.append("title")
    if not excerpt:
        missing.append("excerpt")
    if not date:
        missing.append("date")
    if not author:
        missing.append("author")
    
    return {
        "fields": {"title": bool(title), "excerpt": bool(excerpt), "date": bool(date), "author": bool(author)},
        "missing": missing,
        "pass": len(missing) == 0
    }

# --- Run checks ---
results = []
for post in posts:
    slug = post["slug"]
    raw = post["raw"]
    content = get_content(raw)
    title = get_field(raw, "title")
    
    tfidf = check_tfidf(raw, content)
    entities = check_entities(raw, content)
    pillar = check_pillar_link(raw, content)
    aeo = check_aeo_geo(raw, content)
    internal = check_internal_links(raw, content)
    schema = check_schema(raw, content)
    
    result = {
        "slug": slug,
        "title": title,
        "checks": {
            "TF-IDF": tfidf,
            "Entities": entities,
            "Pillar Link": pillar,
            "AEO/GEO": aeo,
            "Internal Links": internal,
            "Schema Ready": schema,
        }
    }
    results.append(result)

# --- Print Markdown Report ---
pass_count = sum(1 for r in results if all(c["pass"] for c in r["checks"].values()))
fail_count = len(results) - pass_count

now = __import__('datetime').datetime.now().strftime('%Y-%m-%d %H:%M')

print(f"# 📊 Content Framework Audit — kanokmiah.com.bd")
print(f"**Date:** {now} | **Posts checked:** {len(results)}")
print(f"**All checks passed:** {pass_count} ✅ | **Failures:** {fail_count} ❌")
print()

# --- Summary Table ---
print("## 📋 Summary Statistics")
print()
all_tfidf_pass = sum(1 for r in results if r["checks"]["TF-IDF"]["pass"])
all_entities_pass = sum(1 for r in results if r["checks"]["Entities"]["pass"])
all_pillar_pass = sum(1 for r in results if r["checks"]["Pillar Link"]["pass"])
all_aeo_pass = sum(1 for r in results if r["checks"]["AEO/GEO"]["pass"])
all_links_pass = sum(1 for r in results if r["checks"]["Internal Links"]["pass"])
all_schema_pass = sum(1 for r in results if r["checks"]["Schema Ready"]["pass"])

print(f"| Check | ✅ Pass | ❌ Fail | Pass Rate |")
print(f"|-------|--------|--------|-----------|")
total = len(results)
print(f"| TF-IDF Coverage | {all_tfidf_pass} | {total - all_tfidf_pass} | {all_tfidf_pass/total*100:.0f}% |")
print(f"| Entity Coverage | {all_entities_pass} | {total - all_entities_pass} | {all_entities_pass/total*100:.0f}% |")
print(f"| Pillar Link | {all_pillar_pass} | {total - all_pillar_pass} | {all_pillar_pass/total*100:.0f}% |")
print(f"| AEO/GEO Questions | {all_aeo_pass} | {total - all_aeo_pass} | {all_aeo_pass/total*100:.0f}% |")
print(f"| Internal Links | {all_links_pass} | {total - all_links_pass} | {all_links_pass/total*100:.0f}% |")
print(f"| Schema Readiness | {all_schema_pass} | {total - all_schema_pass} | {all_schema_pass/total*100:.0f}% |")
print()

# --- Detail: Failed posts ---
failed_posts = [r for r in results if not all(c["pass"] for c in r["checks"].values())]
if failed_posts:
    print(f"## ❌ Posts Requiring Attention ({len(failed_posts)})")
    print()
    for r in failed_posts:
        print(f"### {r['slug']}")
        print(f"**Title:** {r['title']}")
        print()
        print(f"| Check | Status | Details |")
        print(f"|-------|--------|---------|")
        for ck, cv in r["checks"].items():
            if ck == "TF-IDF":
                kw = cv.get("keyword", "")
                occ = cv.get("occurrences", 0)
                status = "✅" if cv["pass"] else "❌"
                detail = f"Keyword '{kw}' — {occ} occurrences{' (need ≥ 5)' if not cv['pass'] else ''}"
                print(f"| TF-IDF ({kw}) | {status} | {detail} |")
            elif ck == "Entities":
                missing = cv.get("missing_entities", [])
                status = "✅" if cv["pass"] else "❌"
                detail = f"Missing: {', '.join(missing) if missing else 'None'}"
                if cv.get("detected_services"):
                    detail += f" | Services: {', '.join(cv['detected_services'])}"
                print(f"| Entities | {status} | {detail} |")
            elif ck == "Pillar Link":
                status = "✅" if cv["pass"] else "❌"
                detail = f"Links to: {', '.join(cv['pillar_links']) if cv['pillar_links'] else 'None'}"
                print(f"| Pillar Link | {status} | {detail} |")
            elif ck == "AEO/GEO":
                status = "✅" if cv["pass"] else "❌"
                detail = f"{cv['question_heading_count']} question headings{' (need ≥ 2)' if not cv['pass'] else ''}"
                print(f"| AEO/GEO | {status} | {detail} |")
            elif ck == "Internal Links":
                status = "✅" if cv["pass"] else "❌"
                detail = f"{cv['total_internal_links']} total{' (need ≥ 3)' if not cv['pass'] else ''}"
                print(f"| Internal Links | {status} | {detail} |")
            elif ck == "Schema Ready":
                missing = cv.get("missing", [])
                status = "✅" if cv["pass"] else "❌"
                detail = f"Missing: {', '.join(missing) if missing else 'All fields set'}"
                print(f"| Schema Ready | {status} | {detail} |")
        
        print()
        print("**Fix instructions:**")
        for ck, cv in r["checks"].items():
            if not cv["pass"]:
                if ck == "TF-IDF":
                    print(f"- 🔑 **TF-IDF**: Increase usage of keyword '{cv['keyword']}' — currently {cv.get('occurrences', 0)} occurrences, target ≥ 5.")
                elif ck == "Entities":
                    print(f"- 🏷️ **Entities**: Add missing entities: {', '.join(cv.get('missing_entities', []))}")
                elif ck == "Pillar Link":
                    print(f"- 🔗 **Pillar Link**: Add link to /blog/complete-seo-guide-bangladesh-businesses-2026, /services, or /about")
                elif ck == "AEO/GEO":
                    print(f"- 💬 **AEO/GEO**: Add more question-based headings — currently {cv['question_heading_count']}, target ≥ 2")
                elif ck == "Internal Links":
                    print(f"- 🔗 **Internal Links**: Add more internal links — currently {cv['total_internal_links']}, target ≥ 3")
                elif ck == "Schema Ready":
                    print(f"- 📋 **Schema**: Set missing fields: {', '.join(cv.get('missing', []))}")
        print()

# --- Passing posts ---
passed_posts = [r for r in results if all(c["pass"] for c in r["checks"].values())]
if passed_posts:
    print(f"## ✅ Posts Passing All Checks ({len(passed_posts)})")
    for r in passed_posts:
        kw = r["checks"]["TF-IDF"]["keyword"]
        occ = r["checks"]["TF-IDF"]["occurrences"]
        aeo_cnt = r["checks"]["AEO/GEO"]["question_heading_count"]
        links = r["checks"]["Internal Links"]["total_internal_links"]
        print(f"- `{r['slug']}` — kw:'{kw}'({occ}) aeo:{aeo_cnt} links:{links}")

print()
print("---")
print(f"*Report generated by Content Framework Enforcer — {now}*")

File: scripts/framework-audit.mjs
#!/usr/bin/env node
/**
 * Content Framework Enforcer — Blog Post Audit
 * Checks: TF-IDF coverage, Semantic entities, Pillar-cluster alignment,
 *         AEO/GEO optimization, Internal linking, Schema readiness
 */

import { promises as fs } from 'fs';

// ── Helpers ──────────────────────────────────────────────────────────

function extractKeyword(title) {
  // Extract first meaningful noun phrase from title
  const stopwords = new Set(['the', 'a', 'an', 'in', 'for', 'of', 'to', 'and', 'is', 'are', 'what', 'how', 'why', 'when', 'where', 'your', 'our', 'its', 'their', 'complete', 'guide', 'ultimate', 'best', 'top']);
  const words = title.replace(/[^a-zA-Z\s-]/g, '').split(/\s+/).filter(w => w.length > 2 && !stopwords.has(w.toLowerCase()));
  // Also return the full first "key phrase" — first 2-3 meaningful words
  const meaningful = words.filter(w => w.length > 3);
  return meaningful.slice(0, 3).join(' ') || words.slice(0, 3).join(' ') || title.split(' ').slice(0, 3).join(' ');
}

function countOccurrences(content, keyword) {
  if (!keyword || keyword.length < 3) return 999; // skip if keyword too short
  const parts = keyword.toLowerCase().split(' ');
  let count = 0;
  // Check whole keyword phrase
  const phraseRegex = new RegExp(keyword.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'), 'gi');
  count = (content.match(phraseRegex) || []).length;
  // Also check key terms individually  
  for (const part of parts) {
    if (part.length > 3) {
      const termRegex = new RegExp(part.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'), 'gi');
      count += (content.match(termRegex) || []).length;
    }
  }
  return count;
}

function checkEntities(content, title, slug) {
  const requiredEntities = [];
  const lower = content.toLowerCase();
  const titleLower = title.toLowerCase();

  // Location entities
  if (!lower.includes('dhaka') && !slug.includes('dhaka')) {
    requiredEntities.push('Dhaka');
  }
  if (!lower.includes('bangladesh')) {
    requiredEntities.push('Bangladesh');
  }

  // Service type - try to determine from content
  if (lower.includes('seo') || lower.includes('search engine optimization')) {
    if (!lower.includes('seo')) requiredEntities.push('SEO');
  }

  // Author entity
  if (!lower.includes('kanok miah') && !lower.includes('kanok')) {
    // Check if it's a Bangla post
    if (!/[ঀ-৿]/.test(content)) {
      requiredEntities.push('Kanok Miah');
    }
  }

  // Industry/theme specific entities
  if (lower.includes('ecommerce') || lower.includes('e-commerce') || lower.includes('shopify') || lower.includes('daraz')) {
    if (!lower.includes('ecommerce') && !lower.includes('e-commerce')) requiredEntities.push('E-commerce');
  }
  if (lower.includes('garment') || lower.includes('textile') || lower.includes('rmg')) {
    if (!lower.includes('garment')) requiredEntities.push('Garment/Textile');
  }
  if (lower.includes('real estate') || lower.includes('property') || lower.includes('apartment')) {
    if (!lower.includes('real estate')) requiredEntities.push('Real Estate');
  }
  if (lower.includes('mobile') || lower.includes('smartphone') || lower.includes('voice search')) {
    // Common enough, don't flag
  }
  if (lower.includes('google business profile') || lower.includes('gbp') || lower.includes('google maps')) {
    if (!lower.includes('google business profile')) requiredEntities.push('Google Business Profile');
  }
  if (lower.includes('international') || lower.includes('export') || lower.includes('global buyer')) {
    if (!lower.includes('export')) requiredEntities.push('Export');
  }
  if (lower.includes('content market') || lower.includes('content writing')) {
    // common
  }
  if (lower.includes('link building') || lower.includes('backlink')) {
    if (!lower.includes('link building')) requiredEntities.push('Link Building');
  }
  if (lower.includes('keyword research') || lower.includes('keyword')) {
    // common
  }

  return requiredEntities;
}

function countQuestionHeadings(content) {
  const questionStarters = /^#{1,6}\s+(How|What|Why|When|Where|Can|Do|Is|Are|Does|Which|Who)\b/gim;
  const matches = content.match(questionStarters) || [];
  return matches.length;
}

function countInternalLinks(content) {
  // Count links to /blog/, /services/, /locations/, /industries/, /about, /contact, /
  const internalLinkPattern = /\]\(\/(?!http|https|www\.)(?:blog\/|services\/|locations\/|industries\/|about|contact|#)/g;
  const matches = content.match(internalLinkPattern) || [];
  return matches.length;
}

function checkPillarLink(post, allPosts) {
  // Determine pillar topic from tags
  const tags = (post.tags || []).join(' ').toLowerCase();
  let pillarSlug = null;
  let pillarName = null;

  if (tags.includes('seo guide') || tags.includes('bangladesh seo')) {
    pillarSlug = 'complete-seo-guide-bangladesh-businesses-2026';
    pillarName = 'Complete SEO Guide for Bangladesh Businesses 2026';
  } else if (tags.includes('ecommerce') || tags.includes('e-commerce')) {
    pillarSlug = 'why-ecommerce-store-needs-seo-bangladesh';
    pillarName = 'E-commerce SEO guide';
  } else if (tags.includes('technical')) {
    pillarSlug = 'technical-seo-checklist-bangladeshi-websites';
    pillarName = 'Technical SEO guide';
  } else if (tags.includes('local')) {
    pillarSlug = 'google-business-profile-optimization-guide-bangladesh';
    pillarName = 'GBP optimization guide';
  } else if (tags.includes('link building')) {
    pillarSlug = 'link-building-strategies-bangladesh-market';
    pillarName = 'Link Building guide';
  } else if (tags.includes('mobile')) {
    pillarSlug = 'mobile-seo-optimization-bangladesh-mobile-first-era';
    pillarName = 'Mobile SEO guide';
  } else if (tags.includes('content')) {
    pillarSlug = 'content-marketing-strategy-bangladeshi-brands-seo';
    pillarName = 'Content Marketing guide';
  } else if (tags.includes('geo') || tags.includes('aeo') || tags.includes('ai search')) {
    pillarSlug = 'geo-optimization-prepare-business-ai-search';
    pillarName = 'GEO/AEO guide';
  } else if (tags.includes('garment') || tags.includes('textile')) {
    pillarSlug = 'seo-garments-textile-industry-b2b-lead-generation';
    pillarName = 'Garments/Textile SEO guide';
  } else if (tags.includes('real estate')) {
    pillarSlug = 'seo-real-estate-developers-dhaka';
    pillarName = 'Real Estate SEO guide';
  } else if (tags.includes('international') || tags.includes('export')) {
    pillarSlug = 'international-seo-bangladesh-exporters-global-buyers';
    pillarName = 'International SEO guide';
  } else if (tags.includes('semantic') || tags.includes('schema')) {
    pillarSlug = 'schema-markup-rich-snippets-techniques';
    pillarName = 'Schema markup guide';
  } else if (tags.includes('bangla') || tags.includes('বাংলা')) {
    pillarSlug = 'seo-bangla-beginners-guide-google-ranking';
    pillarName = 'SEO Bangla guide';
  } else if (tags.includes('google ads') || tags.includes('ppc')) {
    pillarSlug = 'seo-vs-google-ads-bangladesh-business';
    pillarName = 'SEO vs Google Ads guide';
  }

  if (!pillarSlug) return { ok: null, detail: 'No pillar mapping found for tags: ' + (post.tags || []).join(', ') };

  // Check if pillar is this post itself
  if (post.slug === pillarSlug) return { ok: true, detail: `Self (pillar page: ${pillarName})` };

  // Check content links to pillar
  const contentLower = post.content.toLowerCase();
  const pillarLink = contentLower.includes(pillarSlug) || contentLower.includes(pillarName.toLowerCase().slice(0, 20));
  if (pillarLink) {
    return { ok: true, detail: `Links to pillar: ${pillarName}` };
  }

  return { ok: false, detail: `Missing link to pillar: ${pillarName} (${pillarSlug})` };
}

function checkSchemaReadiness(post) {
  const missing = [];
  if (!post.title || post.title.trim() === '') missing.push('title');
  if (!post.excerpt || post.excerpt.trim() === '') missing.push('excerpt');
  if (!post.date || post.date.trim() === '') missing.push('date');
  if (!post.slug || post.slug.trim() === '') missing.push('slug');
  if (!post.author || post.author.trim() === '') missing.push('author');
  if (missing.length > 0) {
    return { ok: false, detail: `Missing: ${missing.join(', ')}` };
  }
  return { ok: true, detail: 'All basic fields set (title, excerpt, date, slug, author)' };
}

// ── Main ─────────────────────────────────────────────────────────────

async function main() {
  // Read data.js
  const dataJs = await fs.readFile('src/app/blog/data.js', 'utf-8');

  // Extract posts array using eval-safe method
  // Find the posts array definition and reconstruct
  const postsStart = dataJs.indexOf('const posts = [');
  if (postsStart === -1) {
    console.error('Could not find posts array in data.js');
    process.exit(1);
  }

  // We need to evaluate the JS to get the posts. Since it's CommonJS-like, let's
  // write it as a module and import it.
  await fs.writeFile('scripts/_temp_load_posts.mjs', `
    import { createRequire } from 'module';
    const require = createRequire(import.meta.url);
    // Read the file and extract just the posts array
    const fs = require('fs');
    let code = fs.readFileSync('src/app/blog/data.js', 'utf-8');
    // Replace the module.exports line
    code = code.replace('module.exports = posts;', 'export default posts;');
    // Write and import
    const tmpFile = '/tmp/_posts_data.mjs';
    fs.writeFileSync(tmpFile, code);
    const mod = await import(tmpFile);
    export default mod.default;
  `);

  // Actually, let's just use a simpler approach: regex to find each post object
  const posts = [];
  // Find all post objects by splitting on '},' and reconstructing
  // Simpler: just match slug, title, date, excerpt, tags, content fields
  const slugRegex = /slug:\s*"([^"]+)"/g;
  const titleRegex = /title:\s*"([^"]+)"/g;
  const dateRegex = /date:\s*"([^"]+)"/g;
  const excerptRegex = /excerpt:\s*\n\s*"([^"]+)"/g;
  const tagsRegex = /tags:\s*\[([^\]]+)\]/g;
  const contentRegex = /content:\s*`([\s\S]*?)`\s*,/g;

  const slugs = [...dataJs.matchAll(slugRegex)].map(m => m[1]);
  const titles = [...dataJs.matchAll(titleRegex)].map(m => m[1]);
  const dates = [...dataJs.matchAll(dateRegex)].map(m => m[1]);
  const excerpts = [...dataJs.matchAll(excerptRegex)].map(m => m[1]);

  // Parse tags
  const tagMatches = [...dataJs.matchAll(tagsRegex)];
  const tagsList = tagMatches.map(m => {
    const inner = m[1];
    return inner.split(',').map(t => t.trim().replace(/^"|"$/g, '')).filter(t => t);
  });

  // Parse content - this is tricky with backticks
  const contentMatches = [...dataJs.matchAll(/content:\s*`([\s\S]*?)`\s*,?/g)];
  const contents = contentMatches.map(m => m[1]);

  // Parse authors
  const authorRegex = /author:\s*"([^"]+)"/g;
  const authors = [...dataJs.matchAll(authorRegex)].map(m => m[1]);

  // Build posts array
  const maxLen = Math.max(slugs.length, titles.length, dates.length, contents.length);
  for (let i = 0; i < maxLen; i++) {
    posts.push({
      slug: slugs[i] || 'unknown',
      title: titles[i] || '',
      date: dates[i] || '',
      excerpt: excerpts[i] || '',
      tags: tagsList[i] || [],
      author: authors[i] || '',
      content: contents[i] || '',
    });
  }

  // Clean up extracted content - remove any JS code that got mixed in
  for (const post of posts) {
    if (post.content) {
      // Check if the content starts right after backtick, may contain preceding JS
      const contentStart = post.content.search(/^##/m);
      if (contentStart > 0 && contentStart < 100) {
        post.content = post.content.substring(contentStart);
      }
    }
  }

  if (posts.length === 0) {
    console.error('No posts found');
    process.exit(1);
  }

  console.log(`Found ${posts.length} posts total\n`);

  // Track all results
  let allPassed = true;
  let resultsByPost = {};

  for (const post of posts) {
    if (!post.content || post.content.trim() === '') {
      console.log(`## Post: ${post.slug} (${post.title})`);
      console.log('⚠️  Skipped — no content extracted\n');
      continue;
    }

    const keyword = extractKeyword(post.title);
    let keywordDisplay = keyword;

    // Special cases for Bangla posts
    const isBangla = /[\u0980-\u09FF]/.test(post.title);

    // A. TF-IDF Coverage
    const kwCount = countOccurrences(post.content, keyword);
    let tfidfOk = kwCount >= 5 || isBangla; // Bangla posts get pass since keyword extraction is English-focused
    let tfidfDetail = `${kwCount} occurrences`;
    if (!tfidfOk) {
      if (isBangla) {
        tfidfOk = null; // uncertain
        tfidfDetail = `${kwCount} occurrences (Bangla — keyword extraction may be inaccurate)`;
      } else {
        tfidfDetail = `${kwCount} occurrences (needs 5+)`;
      }
    }

    // B. Semantic Entity Coverage
    const missingEntities = checkEntities(post.content, post.title, post.slug);
    let entitiesOk = missingEntities.length === 0;
    let entitiesDetail = entitiesOk ? 'All key entities present' : `Missing: ${missingEntities.join(', ')}`;

    // C. Pillar-Cluster Alignment
    const pillarResult = checkPillarLink(post, posts);
    let pillarOk = pillarResult.ok;
    let pillarDetail = pillarResult.detail;

    // D. AEO/GEO Optimization
    const qHeadings = countQuestionHeadings(post.content);
    let aeoOk = qHeadings >= 2;
    let aeoDetail = `${qHeadings} question headings`;
    if (!aeoOk) aeoDetail += ' (needs 2+)';

    // E. Internal Linking
    const internalLinks = countInternalLinks(post.content);
    let linksOk = internalLinks >= 3;
    let linksDetail = `${internalLinks} total`;
    if (!linksOk) linksDetail += ' (needs 3+)';

    // F. Schema
    const schemaResult = checkSchemaReadiness(post);
    let schemaOk = schemaResult.ok;
    let schemaDetail = schemaResult.detail;

    // ── Report ──
    const checks = {
      'TF-IDF': { ok: tfidfOk, detail: tfidfDetail, key: keywordDisplay },
      'Entities': { ok: entitiesOk, detail: entitiesDetail },
      'Pillar Link': { ok: pillarOk, detail: pillarDetail },
      'AEO/GEO': { ok: aeoOk, detail: aeoDetail },
      'Internal Links': { ok: linksOk, detail: linksDetail },
      'Schema Ready': { ok: schemaOk, detail: schemaDetail },
    };

    resultsByPost[post.slug] = { title: post.title, checks };

    // Determine if post passes overall
    const failures = Object.entries(checks).filter(([_, v]) => v.ok === false);
    const warnings = Object.entries(checks).filter(([_, v]) => v.ok === null);
    const passed = failures.length === 0;

    if (!passed) allPassed = false;
  }

  // ── Generate Report ──
  // Show only posts that have failures (or all if all pass)
  const postsWithIssues = Object.entries(resultsByPost).filter(([_, p]) =>
    Object.values(p.checks).some(c => c.ok === false)
  );
  const postsClean = Object.entries(resultsByPost).filter(([_, p]) =>
    Object.values(p.checks).every(c => c.ok !== false)
  );

  if (postsWithIssues.length === 0) {
    console.log('✅ ALL POSTS PASS — no framework issues found.');
    return;
  }

  for (const [slug, data] of postsWithIssues) {
    console.log(`## Post: ${slug}`);
    console.log(`Title: ${data.title}`);
    console.log('| Check | Status | Details |');
    console.log('|-------|--------|---------|');
    for (const [checkName, checkData] of Object.entries(data.checks)) {
      let status;
      if (checkData.ok === true) status = '✅';
      else if (checkData.ok === null) status = '⚠️';
      else status = '❌';
      const detailKey = checkData.key ? `"${checkData.key}" — ` : '';
      console.log(`| ${checkName} | ${status} | ${detailKey}${checkData.detail} |`);
    }
    console.log('');

    // Fix instructions
    console.log('### Fix instructions:');
    for (const [checkName, checkData] of Object.entries(data.checks)) {
      if (checkData.ok === false) {
        switch (checkName) {
          case 'TF-IDF':
            console.log(`- 🔑 TF-IDF: Increase usage of keyword "${checkData.key}" to at least 5 occurrences across the content.`);
            break;
          case 'Entities':
            console.log(`- 🏷️ Entities: Add mentions of: ${checkData.detail.replace('Missing: ', '')}`);
            break;
          case 'Pillar Link':
            console.log(`- 🔗 Pillar Link: ${checkData.detail}`);
            break;
          case 'AEO/GEO':
            console.log(`- ❓ AEO/GEO: Add question-based H2 headings (How, What, Why, etc.) — need 2+ total.`);
            break;
          case 'Internal Links':
            console.log(`- 🔗 Internal Links: Add more internal links to /blog/, /services/, /locations/, /industries/ — need 3+ total.`);
            break;
          case 'Schema Ready':
            console.log(`- 📋 Schema: ${checkData.detail}`);
            break;
        }
      }
    }
    console.log('');
  }

  // Summary of clean posts
  if (postsClean.length > 0) {
    console.log(`\n✅ ${postsClean.length} posts passed all checks (no issues found).`);
  }
  console.log(`\n📊 Total: ${postsWithIssues.length} posts with issues, ${postsClean.length} clean.`);
}

main().catch(e => {
  console.error('Error:', e);
  process.exit(1);
});
